### DISCUSSION TITLE
Support the equivalent of Autos in VS

### DISCUSSION THREAD

**Discussion by jpwiedekopf** [microsoft/debugpy#1202] [Enhancements]
When debugging a module with many variables, it would be great if the outputs in the debugger could not only be sorted by name, but also by the order of ccess. If this were possible, the last few variables could always be displayed atop the watches.

This would be similar in function to the "Autos" view in Visual Studio, which shows variables used around the current breakpoint: https://docs.microsoft.com/en-us/visualstudio/debugger/autos-and-locals-windows?view=vs-2019

**Comment by int19h** [comment]
It would be nice to have something like VS Autos, but I don't think we can do this from ptvsd end. 

In VS, the way this works is that the language extension (not the debugger) provides a [service](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.textmanager.interop.ivslanguagedebuginfo.getproximityexpressions) that the IDE can use to get the list of variables that are "around" a certain line of code, and the IDE then uses that information to figure out what goes into Autos. The debugger itself isn't even aware of that filter. And there's no equivalent to `GetProximityExpressions` in the VSCode extensibility API (or rather LSP, since logically it would have to be implemented by the language server).

There's also no VSCode UI extensibility point that would allow us to implement custom sorting in Variables pane. That is, we can order them arbitrarily when providing them to the IDE - but there's no way to add a selector to the UI that would allow you to choose whether to order by name or by proximity.

**Comment by jpwiedekopf** [comment]
Thanks very much for the in-depth answer!
This was just something I noticed while debugging some code where I needed a lot of watches to follow what the code is doing, so it would certainly be nice to have. But it's of course no deal-breaker in any way.
Please feel free to close this issue if you feel that is not doable with little effort :)

**Comment by int19h** [comment]
What we might be able to do here is provide a separate "Autos" scope that would function similar to the tool window in VS.  Something like this:
![image](https://user-images.githubusercontent.com/1175142/70743587-ea6edd00-1cd4-11ea-99df-7a0bfa5c3da1.png)
(you'd have to expand it to actually see the variables inside)
